### PR TITLE
WASI: implement argsAlloc and argsFree

### PR DESCRIPTION
**Pull request by shritesh**
Implementing ArgIterator for WASI would require introducing a `free` function to the internally allocated buffer so I've simply introduced a compileError for WASI.

### CODE REVIEW

File: std/os.zig [hunk 1]
```diff
@@ -2166,6 +2170,28 @@ pub fn args() ArgIterator {
 
 /// Caller must call argsFree on result.
 pub fn argsAlloc(allocator: *mem.Allocator) ![]const []u8 {
+    if (builtin.os == Os.wasi) {
+        var count: usize = undefined;
+        var buf_size: usize = undefined;
+
+        _ = os.wasi.args_sizes_get(&count, &buf_size);
```
Comment (andrewrk) [inline std/os.zig]: This should assert that there is no error, or handle it. If it asserts that there is no error, it should explain why the error is impossible.

Comment (shritesh) [reply std/os.zig]: The return values are not documented at all. I looked at the source code of wasmtime and Lucet  and they are not consistent. I have introduced error.ArgsSizesGetFailed and error.ArgsGetFailed in the meantime.

Comment (andrewrk) [reply std/os.zig]: In that case we can follow the pattern set by std/os.zig where an unexpected/undocumented error code is returned. Look for unexpectedErrorPosix and unexpectedErrorWindows.

Generally it's accepted that an undocumented error code might be returned from the OS which translates to error.Unexpected and has a mechanism for logging error codes with a build option.

Comment (shritesh) [reply std/os.zig]: I've changed them to unexpectedErrorPosix. Most of WASI functions (if they are applicable), are similar to POSIX.

File: std/os.zig [hunk 2]
```diff
@@ -2203,6 +2235,16 @@ pub fn argsAlloc(allocator: *mem.Allocator) ![]const []u8 {
 }
 
 pub fn argsFree(allocator: *mem.Allocator, args_alloc: []const []u8) void {
+    if (builtin.os == Os.wasi) {
+        const last_item = args_alloc[args_alloc.len - 1];
+        const last_byte_addr = @ptrToInt(last_item.ptr) + last_item.len + 1; // null terminated
```
Comment (andrewrk) [inline std/os.zig]: If this has a +1 doesn't the other function need a -1?

File: std/os.zig [hunk 3]
```diff
@@ -2134,6 +2134,10 @@ pub const ArgIterator = struct {
     inner: InnerType,
 
     pub fn init() ArgIterator {
+        if (builtin.os == Os.wasi) {
+            @compileError("ArgIterator is not yet supported in WASI. Use argsAlloc and argsFree instead.");
```
Comment (andrewrk) [inline std/os.zig]: Would you mind putting "TODO" in here so that it gets noticed with #363?

File: std/os.zig (hunk 2 above)
Comment (shritesh) [reply std/os.zig]: The strings in `argv_buf`, that is written to by the `args_get` function, are contiguous null terminated c-strings. My assumption here is that `mem.toSlice` doesn't include the final null-terminated buffer and I've added it myself here for the final string.

File: std/os.zig (hunk 3 above)
Comment (shritesh) [reply std/os.zig]: Done

File: std/os.zig (hunk 2 above)
Comment (andrewrk) [reply std/os.zig]: Ah, I see.

**Pull request merged**